src/topology/local_homeomorph.lean

```diff
@@ -606,6 +752,70 @@ local_homeomorph.eq_of_local_equiv_eq $
+`s.ite e.source e'.source = e.source ∩ s ∪ e'.source \ s`, and similarly for target.  The function
+sends `e.source ∩ s` to `e.target ∩ t` using `e` and `e'.source \ s` to `e'.target \ t` using `e'`,
+and similarly for the inverse function. To ensure that the maps `to_fun` and `inv_fun` are inverse
+of each other on the new `source` and `target`, the definition assumes that sets `s` and `t` are
+related both by `e.is_image` and `e'.is_image`. To ensure that the new maps are continuous on
+`source`/`target`, it also assumes that `e.source` and `e'.source` meet `frontier s` on the same set
+and `e x = e' x` on this intersection. -/
+def piecewise (e e' : local_homeomorph α β) (s : set α) (t : set β)
+  [∀ x, decidable (x ∈ s)] [∀ y, decidable (y ∈ t)] (H : e.is_image s t) (H' : e'.is_image s t)
+  (Hs : e.source ∩ frontier s = e'.source ∩ frontier s)
+  (Heq : eq_on e e' (e.source ∩ frontier s)) :
+  local_homeomorph α β :=
+{ to_local_equiv := e.to_local_equiv.piecewise e'.to_local_equiv s t H H',
+  open_source := e.open_source.ite e'.open_source Hs,
+  open_target := e.open_target.ite e'.open_target $
+    H.frontier.inter_eq_of_inter_eq_of_eq_on H'.frontier Hs Heq,
+  continuous_to_fun := continuous_on_piecewise_ite e.continuous_on e'.continuous_on Hs Heq,
+  continuous_inv_fun := continuous_on_piecewise_ite e.continuous_on_symm e'.continuous_on_symm
+    (H.frontier.inter_eq_of_inter_eq_of_eq_on H'.frontier Hs Heq)
+    (H.frontier.symm_eq_on_of_inter_eq_of_eq_on Hs Heq) }
+
+@[simp] lemma piecewise_to_local_equiv (e e' : local_homeomorph α β) {s : set α} {t : set β}
+  [∀ x, decidable (x ∈ s)] [∀ y, decidable (y ∈ t)] (H : e.is_image s t) (H' : e'.is_image s t)
+  (Hs : e.source ∩ frontier s = e'.source ∩ frontier s)
+  (Heq : eq_on e e' (e.source ∩ frontier s)) :
+  (e.piecewise e' s t H H' Hs Heq).to_local_equiv =
+    e.to_local_equiv.piecewise e'.to_local_equiv s t H H' :=
+rfl
+
+@[simp] lemma piecewise_coe (e e' : local_homeomorph α β) {s : set α} {t : set β}
+  [∀ x, decidable (x ∈ s)] [∀ y, decidable (y ∈ t)] (H : e.is_image s t) (H' : e'.is_image s t)
+  (Hs : e.source ∩ frontier s = e'.source ∩ frontier s)
+  (Heq : eq_on e e' (e.source ∩ frontier s)) :
+  ⇑(e.piecewise e' s t H H' Hs Heq) = s.piecewise e e' :=
+rfl
+
+@[simp] lemma symm_piecewise (e e' : local_homeomorph α β) {s : set α} {t : set β}
+  [∀ x, decidable (x ∈ s)] [∀ y, decidable (y ∈ t)] (H : e.is_image s t) (H' : e'.is_image s t)
+  (Hs : e.source ∩ frontier s = e'.source ∩ frontier s)
+  (Heq : eq_on e e' (e.source ∩ frontier s)) :
+  (e.piecewise e' s t H H' Hs Heq).symm =
+    e.symm.piecewise e'.symm t s H.symm H'.symm
+      (H.frontier.inter_eq_of_inter_eq_of_eq_on H'.frontier Hs Heq)
+      (H.frontier.symm_eq_on_of_inter_eq_of_eq_on Hs Heq) :=
+rfl
+
+/-- Combine two `local_homeomorph`s with disjoint sources and disjoint targets. We do not reuse
+`local_homeomorph.piecewise` here to provide better formulas for `source` and `target`. -/
```

Comment:
The docstring is misleading, since you *do* reuse `local_homeomorph.piecewise`, albeit modifying it afterwards for better defeq behavior.